spanner/session.go

```diff
@@ -1152,8 +1152,10 @@ func (p *sessionPool) remove(s *session, isExpire bool) bool {
 	if s.invalidate() {
 		// Decrease the number of opened sessions.
 		p.numOpened--
-		// Decrease the number of sessions in use.
-		p.decNumInUseLocked(ctx)
+		// Decrease the number of sessions in use, only when not from idle list.
+		if !isExpire {
```

Comment:
I think that it would be better to add an additional argument to the `remove(..)` method that explicitly says whether the session was in use or not. So something like this:

```go
func (p *sessionPool) remove(s *session, isExpire bool, wasInUse bool) {
  ...
  if wasInUse {
    p.decNumInUseLocked(ctx)
  }
}
```

In theory, it could be that this method is called in the future to remove sessions that have not expired, but that also were not in use at the time that they were being removed, and then we could re-introduce a similar bug as the one here. That is less likely with an explicit argument that clearly says what it is for.